### ISSUE TITLE
Is SSHFS supported?

### ISSUE THREAD

**Issue by vansante** [pkg/sftp#265]
It seems that SSHFS is not supported, because mounting the SFTP server with SSHFS works, and most things will actually work, except for file creation, uploading, etc. Is this intended? Or is pkg/sftp supposed to support SSHFS?

The steps I have taken to test this are as follows:

- Start a test SFTP server with the example at `https://github.com/pkg/sftp/tree/master/examples/sftp-server`
- Mount it with SSHFS like so: 
```
paul@paul-VirtualBox:~$ sshfs -p 2022 "testuser"@192.168.56.1:/ /home/paul/sshfs
testuser@192.168.56.1's password: 
```
- And then attempt to copy a few pictures like so:
```
paul@paul-VirtualBox:~$ cp plaatjes/*.png sshfs/mnt/d/sftp/
cp: cannot create regular file 'sshfs/mnt/d/sftp/5136c62b07be1.png': No such file or directory
cp: cannot create regular file 'sshfs/mnt/d/sftp/Azure-DNS.png': No such file or directory
cp: cannot create regular file 'sshfs/mnt/d/sftp/moreinfogone.png': No such file or directory
```

Some pictures will actually copy, but generally there will always be a large amount that fail. (The randomness seems to me like a timing issue perhaps)

Is this a bug or is SSHFS something the package just doesn't support?

**Comment by puellanivis** [comment]
I don’t think it was thought about as a support point before. But, I think the use case is intended to be targeting SFTP clients specifically, rather than providing the complete generic VFS system itself.

**Comment by eikenb** [comment]
I think I've figured it out. It is the damn packet ordering again. The packet order code has 2 modes; a command mode where it has 1 worker and processes commands sequentially and a read/write mode where it has a pool of workers for handling file reads and writes faster. The mode starts in the command mode and switches to read/write mode when an Open packet is received (and switched back on the Close packet).

In this case the client (sshfs) is sending the Open packet for a file immediately followed by a Stat packet. The Open packet flips it to r/w mode and passes the Open packet to a worker. The Stat packet comes in then and is given to another worker... the Stat then gets checked before the Open runs and finds no file yet.
When that happens (not every time due to it being a race) you get the file not found error (from the Stat) w/ the file created (from the Open) but with no content (as the client stops the write due to the error).

Thinking about the best way to fix. Initial idea is to look into changing the logic around flipping into R/W mode to flip based on seeing the (file) Handle packet packet (the response to the Open packet). But I'm not sure how to do this nicely right off.

**Comment by puellanivis** [comment]
Ugh… yeah. Looks like a fencing issue, where when transitioning between states a fence need be written… but of course this is easier said than done.

**Comment by eikenb** [comment]
I think I have a good idea for a fix. Instead Open/Close packets switching state, ignore the Open and just send Read/Write packets to the worker pool and everything else to the single worker. Keeping the wg.Wait on the Close to make sure all RW is complete.

I'll post a PR with the change once I'm happy with it and come up with a test (if I can). I was thinking about how to test it (how to insert the delay to trigger the race) when I came up with the above.

**Comment by eikenb** [comment]
I spent much of the day trying to think of a good way to inject the delay to trigger the race but have given up for now. I left in the test with a comment about this and about how I want to add a way to inject delays later.

Pull request should be up in a couple.

**Comment by eikenb** [comment]
PR #266 is up w/ my fix.

**Comment by vansante** [comment]
I have done a little test with your issue branch and it seems to work on my end :)

Thanks for working on this!
